src/rdb/rdb_raft.c

```diff
@@ -1119,54 +1119,139 @@ rdb_raft_update_node(struct rdb *db, uint64_t index, raft_entry_t *entry, rdb_vo
+		DL_ERROR(rc, DF_DB ": failed to begin VOS TX for entry %ld", DP_DB(db), index);
+		return rc;
+	}
 
 	/*
-	 * If this is an rdb_tx entry, apply it. Note that the updates involved
-	 * won't become visible to queries until entry index is committed.
-	 * (Implicit queries resulted from rdb_kvs cache lookups won't happen
-	 * until the TX releases the locks for the updates after the
-	 * rdb_tx_commit() call returns.)
+	 * Update the log tail. To get the same minor epoch for all MC updates
+	 * across different VOS TXs, we update the MC first.
 	 */
+	D_ASSERTF(index == db->d_lc_record.dlr_tail, DF_U64 " == " DF_U64 "\n", index,
+		  db->d_lc_record.dlr_tail);
+	db->d_lc_record.dlr_tail = index + 1;
+	dirtied_tail = true;
+	d_iov_set(&values[0], &db->d_lc_record, sizeof(db->d_lc_record));
+	rc = rdb_mc_update(db->d_mc, RDB_MC_ATTRS, 1 /* n */, &rdb_mc_lc, &values[0], vtx);
+	if (rc != 0) {
+		DL_ERROR(rc, DF_DB ": failed to update log tail " DF_U64, DP_DB(db),
+			 db->d_lc_record.dlr_tail);
+		goto out_vtx;
+	}
+
 	if (entry->type == RAFT_LOGTYPE_NORMAL) {
-		rc = rdb_tx_apply(db, index, entry->data.buf, entry->data.len,
-				  rdb_raft_lookup_result(db, index), &crit, vtx);
-		if (rc != 0) {
-			DL_ERROR(rc, DF_DB ": failed to apply entry " DF_U64, DP_DB(db), index);
-			return rc;
+		if (!skip_tx_apply) {
+			/*
+			 * If this is an rdb_tx entry, apply it. Note that the updates involved
+			 * won't become visible to queries until entry index is committed.
+			 * (Implicit queries resulted from rdb_kvs cache lookups won't happen
+			 * until the TX releases the locks for the updates after the
+			 * rdb_tx_commit() call returns.)
+			 */
+			rc = rdb_tx_apply(db, index, entry->data.buf, entry->data.len,
+					  rdb_raft_lookup_result(db, index), &crit, vtx);
+			if (rc == RDB_TX_APPLY_ERR_DETERMINISTIC) {
+				/*
+				 * We must abort VOS TX to discard any partial application of the
+				 * entry, and begin a new VOS TX to store the entry without applying
+				 * it. The new VOS TX will reuse crit.
+				 */
+				D_DEBUG(DB_TRACE, DF_DB ": deterministic error for entry %ld\n",
+					DP_DB(db), index);
+				rc = -DER_AGAIN;
+				skip_tx_apply = true;
+				goto out_vtx;
+			} else if (rc != 0) {
+				DL_ERROR(rc, DF_DB ": failed to apply entry " DF_U64, DP_DB(db),
+					 index);
+				goto out_vtx;
+			}
+			dirtied_kvss = true;
 		}
 	} else if (raft_entry_is_cfg_change(entry)) {
-		crit = true;
 		rc = rdb_raft_update_node(db, index, entry, vtx);
 		if (rc != 0) {
 			DL_ERROR(rc, DF_DB ": failed to update replicas " DF_U64, DP_DB(db), index);
-			return rc;
+			goto out_vtx;
 		}
 	} else {
-		D_ERROR(DF_DB": unknown entry "DF_U64" type: %d\n", DP_DB(db), index, entry->type);
-		return -DER_IO;
+		D_ERROR(DF_DB ": unknown entry " DF_U64 " type: %d\n", DP_DB(db), index,
+			entry->type);
+		rc = -DER_IO;
+		goto out_vtx;
 	}
 
 	/*
 	 * Persist the header and the data (if nonempty). Discard the unused
 	 * entry->id.
 	 */
+	n = 0;
```

Comment:
[Note] This change is unnecessary, but since we now retry in this function, moving `n = 0` closer to `n`'s usage is clearer.